File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        # A set provides O(1) instant lookups to see if a character is in our window
        char_set = set()
        left = 0
        max_length = 0
        
        for right in range(len(s)):
            # If we hit a duplicate, shrink the window from the left 
            # until the duplicate is no longer inside
            while s[right] in char_set:
                char_set.remove(s[left])
                left += 1
            
            # Add the current valid character to our window's memory
            char_set.add(s[right])
            
            # Record the length of the window if it's our new high score
            # The length of the window is simply (right pointer - left pointer + 1)
            max_length = max(max_length, right - left + 1)
            
        return max_length

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
import java.util.HashSet;
import java.util.Set;

class Solution {
    public int lengthOfLongestSubstring(String s) {
        Set<Character> set = new HashSet<>();
        int left = 0;
        int maxLength = 0;

        for (int right = 0; right < s.length(); right++) {
            // Shrink window from the left until duplicate character is removed
            while (set.contains(s.charAt(right))) {
                set.remove(s.charAt(left));
                left++;
            }
            
            // Add current character to window set and update maximum length
            set.add(s.charAt(right));
            maxLength = Math.max(maxLength, right - left + 1);
        }

        return maxLength;
    }
}